SIMP Optimization

recuero opened the pull request:
Closes #25602

maxnezdyur committed:
initial SIMP formulation

maxnezdyur committed:
add documentation

maxnezdyur committed:
add top opt tests

maxnezdyur committed:
get rid fo filtered density and clean up code

maxnezdyur committed:
fix tests to get rid of filtered density

maxnezdyur committed:
mesh size insensitive formulation

maxnezdyur committed:
pde hemoltz filter changes

maxnezdyur committed:
fix tests and add comments and add license

maxnezdyur committed:
add more docs and fix mainpage

maxnezdyur committed:
webpage setup

maxnezdyur committed:
save website info

maxnezdyur committed:
add tests and fix folder setup

maxnezdyur committed:
finish examples documentation

maxnezdyur committed:
address comments and update to simpler syntax

lynnmunday committed:
compliance sensitivity computed with a derivative parsed material closes #3

recuero committed:
Temporary commit with cost sensitivity

recuero committed:
Migrate existing tests to derivative parsed materials (#5)

recuero committed:
Delete old version of compliance sensitivity

lynnmunday committed:
non working ordered multimaterial simp

recuero committed:
Temp commit with cost.

recuero committed:
Preliminary multimaterial SIMP implementation

lynnmunday committed:
input file and python plotting script to check parsedMaterial is giving the correct density dependent youngs modulus

recuero committed:
Initial verification of multimaterial density-based optimization.

For converged results, increase the test's number of allowed iterations.

We should create an example out of this.

recuero committed:
Add example with adaptive move.

Adaptive move is dangerous. It will greatly move the solution toward
particular minima, including single material solution. Use with caution.

lynnmunday committed:
added some outputs and changed tests to do 3 materials

lynnmunday committed:
adding objective and constraint plotting script along with input file to create the data.

recuero committed:
Output correct total cost for multimaterial SIMP

recuero committed:
Initial stab at thermal compliance minimization.

Results "look" consistent with the literature. But more testing is needed.
The objective function grows after a few iterations, which may point to an
implementation of the thermal sensitivity that is not completely correct.

To be reviewed.

lynnmunday committed:
single material thermal compliance minimization

recuero committed:
Consider block restriction in optimization user objects.

I am also regolding a test, but this is unrelated with the
changes in this commit. I should have regolded it a few commits
ago.

recuero committed:
Add multimaterial examples.

It includes:

 - Multiblock with single materials, all optimized
 - Multiblock with single materials, all but one optimized

recuero committed:
Add two multi-loads examples.

These examples show that:

 - Accounting for loads in slightly different locations change drastically
   the topology of the optimized structure.
 - Accounting for nonsimultaneous loading, when the loads have a null resulting
   effect if considered together, is necessary to achieve a nontrivial solution.

recuero committed:
Add thermal compliance test

GiudGiud commented:
SIMP?

recuero commented:
Solid Isotropic Material Penalization (spelled out in the issue)

GiudGiud commented:
Should we add a letter? To make it pc. SOlid Isotropic Material Penalization ?

recuero commented:
We could, I'd be fine with it. But it may make it challenging for users to identify what it is that we are doing.

recuero commented:
https://en.wikipedia.org/w/index.php?title=Topology_optimization&action=edit&section=4

lynnmunday commented:
It has to stay as simp, that is what everyone calls it.

recuero committed:
Move files (#25602)

Move files to optimization, tensor mechanics, heat conduction,
and combined modules.

Regold tests that should have been regolded with updated computation
of thermal compliance sensitivity.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1796023/) on 620f7aa wanted to post the following:

Warning: This PR changes repo size by 13.75 MiB.

GiudGiud commented on modules/combined/doc/content/modules/combined/examples/optimization/images/2d_mbb.mp4:
these have to go in large_media

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1796019/) on 620f7aa wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25603/site)

This comment will be updated on new commits.

recuero committed:
Fix SQA.

This commits contains changes to:

 - Add documentation
 - Make tests pass and not time out
 - Ensure coverage

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 3cd818f6f1ab70dd30c7d68c8aea6ba6f5251a6b
+Subproject commit fe113cb23869a0c223502fae357bf690b7881d6a
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1798652/) on 3beb0fd wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>3923fc</th>
      <th colspan=4>#25603 3beb0f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.69%</td>
      <td>85.69%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>91789</td>
      <td>91811</td>
      <td>+22</td>
      <td>22</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15330</td>
      <td>15330</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25603/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25603/coverage/framework)


## Modules coverage

### Heat conduction

<table>
  <thead>
    <tr>
      <th></th>
      <th>3923fc</th>
      <th colspan=4>#25603 3beb0f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.46%</td>
      <td>87.72%</td>
      <td>-0.74%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3679</td>
      <td>3679</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>480</td>
      <td>515</td>
      <td>+35</td>
      <td>35</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25603/coverage_diff/heat_conduction)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25603/coverage/heat_conduction)

### Optimization

<table>
  <thead>
    <tr>
      <th></th>
      <th>3923fc</th>
      <th colspan=4>#25603 3beb0f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.99%</td>
      <td>76.31%</td>
      <td>-12.68%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>1366</td>
      <td>1366</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>169</td>
      <td>424</td>
      <td>+255</td>
      <td>255</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25603/coverage_diff/optimization)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25603/coverage/optimization)

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>3923fc</th>
      <th colspan=4>#25603 3beb0f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.63%</td>
      <td>84.59%</td>
      <td>-0.04%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>25763</td>
      <td>25763</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>4679</td>
      <td>4695</td>
      <td>+16</td>
      <td>16</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25603/coverage_diff/tensor_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25603/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25603/coverage/xfem)

</details>

## Warnings

- `heat_conduction` new line coverage rate 0.00% is less than the suggested 90.0%
- `optimization` new line coverage rate 0.00% is less than the suggested 90.0%
- `tensor_mechanics` new line coverage rate 0.00% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1798652/) on 3beb0fd wanted to post the following:


The following coverage requirement(s) failed:
- `optimization` coverage rate 76.31% is less than the required 86.0%

recuero committed:
Cover code within their own modules.

recuero committed:
Add mult-app thermomechanical SIMP example.

Also add more documentation to other examples.

recuero committed:
Last touches before review

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 3cd818f6f1ab70dd30c7d68c8aea6ba6f5251a6b
+Subproject commit 7f88f52ee687c3d09b51bfb6f4ccdfaa0fd6ba79
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

recuero committed:
Remove unnecessary files, move optimization tests, fix location tests.

recuero commented on modules/optimization/src/userobjects/SensitivityFilter.C:
```diff
@@ -0,0 +1,65 @@
+  // Find the current element in the filter
+  auto filter_iter = _filter.find(_current_elem->id());
+
+  // If the element is not found in the filter, throw an error
```
assert

recuero commented on modules/combined/test/tests/optimization/compliance_sensitivity/2d_mmb_2material.i:
```diff
@@ -0,0 +1,204 @@
+    property_name = E_phys
+  []
+  # [E_phys]
+  #   type = DerivativeParsedMaterial
```
remove

recuero commented on modules/combined/test/tests/optimization/compliance_sensitivity/2d_mmb_2material_cost.i:
```diff
@@ -0,0 +1,217 @@
+  #   type = DerivativeParsedMaterial
+  #   # Emin + (density^penal) * (E0 - Emin)
+  #   expression = 'A1:=(${E0}-${E1})/(${rho0}^${power}-${rho1}^${power}); B1:=${E0}-A1*${rho0}^${power}; A1*mat_den^${power}+B1'
+  #   coupled_variables = 'mat_den'
```
remove

recuero commented on modules/combined/examples/optimization/helmholtz_multimat_nostrip.i:
```diff
@@ -0,0 +1,367 @@
+
+# [ICs]
+#   [mat_den]
+#     type = RandomIC
```
remove

recuero commented on modules/combined/test/tests/optimization/compliance_sensitivity/tests:
```diff
@@ -0,0 +1,70 @@
+    heavy = true
+  []
+  [3d_mbb]
+    type = 'CSVDiff'
```
Run with one core too.

recuero committed:
Add doxygen comments

lindsayad commented on framework/include/bcs/ADRobinBC.h:
```diff
@@ -20,4 +20,7 @@ class ADRobinBC : public ADIntegratedBC
 protected:
   ADReal computeQpResidual() override;
+
+  // Coefficient to be applied to the Robin boundary condition
```
```suggestion
  /// Coefficient to be applied to the Robin boundary condition
```

lindsayad commented on framework/include/bcs/ADVectorRobinBC.h:
```diff
@@ -20,4 +20,7 @@ class ADVectorRobinBC : public ADVectorIntegratedBC
 protected:
   ADReal computeQpResidual() override;
+
+  // Coefficient to be applied to the Robin boundary condition
```
```suggestion
  /// Coefficient to be applied to the Robin boundary condition
```

lindsayad reviewed (commented):
Just reviewing the framework changes

lindsayad commented:
@lynnmunday this seems like a good PR for you to review

recuero committed:
Address Alex and Max's reviews

Also simplify some logic.

maxnezdyur commented on modules/optimization/include/userobjects/DensityUpdateTwoConstraints.h:
```diff
@@ -0,0 +1,119 @@
+
+/**
+ * Element user object that performs SIMP optimization using a bisection algorithm applying a volume
+ * constraint and a cost constaint.
```
```suggestion
 * constraint and a cost constraint.
```

maxnezdyur commented on modules/optimization/include/userobjects/DensityUpdateTwoConstraints.h:
```diff
@@ -0,0 +1,119 @@
+   */
+  void gatherElementData();
+  /**
+   * Performs the optimility criterion loop (bisection)
```
```suggestion
   * Performs the optimality criterion loop (bisection)
```

maxnezdyur commented on modules/optimization/include/userobjects/DensityUpdate.h:
```diff
@@ -0,0 +1,82 @@
+  virtual void threadJoin(const UserObject &) override{};
+
+protected:
+  /// The sytem mesh
```
```suggestion
  /// The system mesh
```

maxnezdyur commented on modules/optimization/include/userobjects/DensityUpdate.h:
```diff
@@ -0,0 +1,82 @@
+  void gatherElementData();
+
+  /**
+   * Performs the optimility criterion loop (bisection)
```
```suggestion
   * Performs the optimality criterion loop (bisection)
```

maxnezdyur commented on modules/optimization/doc/content/modules/optimization/examples/index.md:
```diff
@@ -10,3 +10,8 @@ All of the examples in this section are parameterizing force loads or material p
+
+Solid Isotropic Material Penalization (SIMP) examples typically require mechanical or thermal physics to meaningfully leverage
+the optimization module's capabilities. Some SIMP examples, including those that combine multiple loads, multiple materials, and
+multiple physics are availble in the MOOSE combined module.
```
```suggestion
multiple physics are available in the MOOSE combined module.
```

maxnezdyur commented on modules/heat_conduction/include/materials/ThermalSensitivity.h:
```diff
@@ -0,0 +1,46 @@
+
+/**
+ * Computes heat conduction compliance derivative with respect to a variable contained in a parsed
+ * material. This material can be used to optimizate a system thermal compliance objective function
```
```suggestion
 * material. This material can be used to optimize a system thermal compliance objective function
```

maxnezdyur commented on modules/optimization/src/userobjects/DensityUpdateTwoConstraints.C:
```diff
@@ -0,0 +1,300 @@
+void
+DensityUpdateTwoConstraints::gatherElementData()
+{
+  // Create parallel-consistent data structures constaining compliance and cost sensitivities.
```
```suggestion
  // Create parallel-consistent data structures constraining compliance and cost sensitivities.
```

maxnezdyur reviewed (commented):
Great documentation, mostly just found some spelling mistakes. We talked about the compliance  sensitivity and possibly the need to divide by the effective young modulus, only major thing that we need to agree on.

recuero commented on modules/tensor_mechanics/src/materials/ComplianceSensitivity.C:
```diff
@@ -0,0 +1,42 @@
+{
+  // Call the parent class's method to compute the strain energy density
+  StrainEnergyDensity::computeQpProperties();
+  _sensitivity[_qp] = -_dEdp[_qp] * _strain_energy_density[_qp];
```
Yes, we do. I went back and forth testing that, after our conversation, and somehow ended up with the wrong code.

recuero commented on modules/optimization/src/userobjects/DensityUpdateTwoConstraints.C:
```diff
@@ -0,0 +1,300 @@
+  // Control adaptivity
+  const Real alpha = 0.96;
+
+  if (_adaptive_move)
+  {
+    auto transient = dynamic_cast<Transient *>(_app.getExecutioner());
+    if (transient)
+      move = std::max(MathUtils::pow(alpha, transient->getTime()) * move, move_min);
+    else
+      mooseError("Cannot find a transient executioner for adaptive bisection move.");
+  }
```
You are right that time is in the user object. But this flavor of "move" relies on a transient executioner, so I think it's good to check that we are in a transient type of execution.

recuero replied:
By "rely" I mean it doesn't make sense otherwise.

recuero commented on modules/optimization/src/userobjects/DensityUpdateTwoConstraints.C:
```diff
@@ -0,0 +1,300 @@
+  const Real damping = 1.0;
+
+  Real updated_density = std::max(
+      1.0e-5,
```
It is a small "delta" that some authors use. But there may be no need. I replaced it with zero and will regold accordingly.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1825072/) on d0db252 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25603/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25603/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6d70fba25568259d2b64e247e42717a33ce7778a`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1825504/) on 923065b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25603/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25603/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6d70fba25568259d2b64e247e42717a33ce7778a`

lindsayad commented:
@recuero please mention me when you are done with @maxnezdyur's review

lynnmunday commented on modules/optimization/src/userobjects/DensityUpdateTwoConstraints.C:
```diff
@@ -0,0 +1,292 @@
+  params.addClassDescription(
+      "Compute updated densities based on sensitivities using an optimality criteria method to "
+      "keep the volume and cost constraints satisified.");
+  params.addRequiredParam<Real>("power", "Penalty power for SIMP method.");
```
```suggestion
```

lynnmunday commented on modules/optimization/src/userobjects/DensityUpdateTwoConstraints.C:
```diff
@@ -0,0 +1,292 @@
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "DensityUpdateTwoConstraints.h"
```
A future refactor should try to create a base class to share common functions with DensityUpdate.

recuero replied:
Yes. That is a byproduct of not wanting to touch Max's initial implementation. That can be future work.

lynnmunday commented on modules/optimization/src/userobjects/DensityUpdateTwoConstraints.C:
```diff
@@ -0,0 +1,292 @@
+    // Loop over all elements
+    for (auto && [id, elem_data] : _elem_data_map)
+    {
+      const std::vector<Real> & weight_values =
```
Should weight_values be member data filled in the constructor?

lynnmunday reviewed (commented):
I can make these changes if you comment on them.

recuero commented on modules/combined/doc/content/modules/combined/examples/optimization/2d_mbb_pde.md:
```diff
@@ -0,0 +1,45 @@
+
+One advantage of using a PDE filter is that by applying boundary conditions
+on the sensitivity variable on the boundary of the domain the filter will
+prevent "sticking" of the material commonly seen in topology optimization. That
```
There isn't a bunch of figures in large media. There are a few of them. There are linked whenever they were supposed to be linked.

lynnmunday replied:
ok.  I'll add links to other figures in a separate commit.

recuero replied:
That figure is already in large_media.

recuero commented on modules/combined/doc/content/modules/combined/examples/optimization/2d_mbb_pde_amr.md:
````diff
@@ -0,0 +1,33 @@
+The main addition for this example is showing an effective indicator for
+adaptive mesh refinement, which is the `mat_den_nodal` variable below.. In the `AuxVariables` block there is a new variable.
+
+```
````
Max had it that way. I can make that link.

recuero commented on modules/combined/doc/content/modules/combined/examples/optimization/images/2d_mbb_pde_amr.png:
That comment is "outdated". You reviewed old code.

recuero commented on modules/combined/doc/content/modules/combined/examples/optimization/multiload.md:
```diff
@@ -0,0 +1,44 @@
+
+We employ here a multi-app scheme similar to that employed in
+[Thermal and mechanical optimization](combined/examples/optimization/thermomechanical.md).
+In this case, however, we solve similar physics in the subapps. In fact, there is only a slight
+difference between: The single load in each of them are located at different material
+points. Considering the optimization of a problem with the separate effect of the loads, instead
+of accounting for its simultaneous application, can yield a final optimized system whose
+material distribution is a lot different from a simple concomitant approach. In the extreme case,
+having two loads acting on the same point in opposing directions would yield an ill-defined
+optimized structure if such loads are considered to be concomitant. However, the independent
+combination of each of the load's sensitivities would yield a structure that is optimized to
+reduce the compliance under such weighting.
```
> ielding an optimized structure that is minimally compliant under each load

That is not correct. It's the response to the combination of the loads that's minimally compliant. I can rephrase that.

recuero replied:
> Optimizing the structure with separately applied loads is preferable to simultaneously applied loads

This also depends... I wouldn't say so in general. 

I'll try to include some of your comments here.

recuero replied:
Mentioning here the weighting of sensitivities is relevant.

recuero commented on modules/heat_conduction/test/tests/thermal_materials/2d.i:
```diff
@@ -0,0 +1,142 @@
+[]
+
+[Materials]
+  [thermal_compliance]
+    type = ThermalCompliance
+    temperature = u
+    thermal_conductivity = thermal_cond
+    outputs = 'exodus'
+  []
+  [thermal_cond]
+    type = DerivativeParsedMaterial
+    expression = "A1:=(${TC0}-${TC1})/(${rho0}^${power}-${rho1}^${power}); "
+                 "B1:=${TC0}-A1*${rho0}^${power}; TC1:=A1*mat_den^${power}+B1; TC1"
+    coupled_variables = 'mat_den'
+    property_name = thermal_cond
+    outputs = 'exodus'
+  []
+  [thermal_compliance_sensitivity]
+    type = ThermalSensitivity
+    design_density = mat_den
+    thermal_conductivity = thermal_cond
+    temperature = u
+    outputs = 'exodus'
+  []
```
Yes. We have to for coverage purposes.

recuero commented on modules/combined/test/tests/optimization/compliance_sensitivity/3d_mbb.i:
```diff
@@ -0,0 +1,238 @@
+  []
+[]
+
+[Controls]
```
@maxnezdyur

maxnezdyur replied:
Applying the RobinBC on all edges, then turning off the ones you don't want it on has a quicker convergence rate than not doing it. It only gave enough of a speedup on the 3D cases, so I had it so I wouldn't timeout a CIVET test.

recuero commented on modules/combined/test/tests/optimization/compliance_sensitivity/2d_mmb_2material_cost.i:
```diff
@@ -0,0 +1,211 @@
+  petsc_options_value = 'lu superlu_dist'
+  nl_abs_tol = 1e-8
+  dt = 1.0
+  num_steps = 70
```
Those tests run as 'heavy'. The original author, Max, set them to 70 steps. We need to have a reason to change it to 10.

maxnezdyur replied:
This is set up to 70 because at 10 time steps the result was not optimized enough to resemble a structure.

recuero replied:
That's good to have more "demonstation-like" tests. They are marked as heavy so they don't generate coverage, they also don't run in debug mode and are marked not to undergo valgrind testing.

Having shorter-running/smaller tests is also convenient to exercise the logic and not having to constrain it to 'heavy' testing. We should aim at having most tests run quickly (< 1 or 2 seconds).

recuero commented on modules/combined/test/tests/optimization/compliance_sensitivity/2d_mmb_2material_cost.i:
```diff
@@ -0,0 +1,211 @@
+[]
+
+[Postprocessors]
+  [total_vol]
```
Yes, that can be done in your follow-on PR.

recuero commented on modules/combined/test/tests/optimization/compliance_sensitivity/2d_mmb_2material_cost.i:
```diff
@@ -0,0 +1,211 @@
+[UserObjects]
+  [rad_avg]
+    type = RadialAverage
+    radius = 1.2
```
That depends on the case, just like in the literature. But typically it should go from a little more than one element to 4 or 5 elements.

recuero commented on modules/combined/examples/optimization/three_materials.i:
```diff
@@ -0,0 +1,312 @@
+    bisection_lower_bound = 0
+    bisection_upper_bound = 1.0e16 # 100
+
+    # moves of 0.5 typically drive solution to "black and white"
```
I'll remove the comment.

maxnezdyur commented on modules/combined/doc/content/modules/combined/examples/optimization/2d_mbb.md:
```diff
@@ -0,0 +1,52 @@
+# 2D MBB Beam with a Convolution Filter
+
+In this example we will go through the general setup of a topology optimization
```
@article{article,
author = {Sigmund, Ole},
year = {2007},
month = {04},
pages = {401-424},
title = {Morphology-based black and white filters for topology optimization},
volume = {33},
journal = {Structural and Multidisciplinary Optimization},
doi = {10.1007/s00158-006-0087-x}
}
This is the paper on filtering that I used most.

recuero commented on modules/combined/doc/content/modules/combined/examples/optimization/multiload.md:
```diff
@@ -0,0 +1,44 @@
+# Multiple Load SIMP
```
Yes, that's the 2003 book by Bendsoe and Sigmund.

maxnezdyur commented on modules/combined/doc/content/modules/combined/tutorials/index.md:
```diff
@@ -5,3 +5,21 @@ physics modules can be used together:
+various conditions, whether they be various loads acting on a single system or various physics
+acting on the same mesh domain.
+
+All the following examples employ SIMP topology optimization:
```
Posted in another response I would cite this one also,
@article{filtering,
author = {Sigmund, Ole},
year = {2007},
month = {04},
pages = {401-424},
title = {Morphology-based black and white filters for topology optimization},
volume = {33},
journal = {Structural and Multidisciplinary Optimization},
doi = {10.1007/s00158-006-0087-x}
}

recuero commented on modules/combined/test/tests/optimization/compliance_sensitivity/3d_mbb.i:
```diff
@@ -0,0 +1,238 @@
+  print_linear_residuals = false
+[]
+
+[Postprocessors]
```
While not optimal, the postprocessors are good to avoid regression. I suggest you create a follow-on PR when you decide which specific postprocessors you'd like in your optimization inputs.

recuero commented on modules/optimization/doc/content/source/userobjects/DensityUpdateTwoConstraints.md:
```diff
@@ -0,0 +1,26 @@
+
+## Description
+
+The `DensityUpdateTwoConstraints` class is an `ElementUserObject`
```
This methodology corresponds to a multimaterial paper (Multi-material topology optimization using ordered SIMP
interpolation)

recuero committed:
Address Lynn's comments

recuero commented:
@lynnmunday I have address the vast majority of your comments. Those related to having a particular number of time steps or specific postprocessors remain as future work. Those can be addressed in a follow-on PR when the optimization team agrees on the right postprocessors for regression testing.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1827561/) on 93f9e6f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25603/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25603/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 723a2b6ba9a9919726fe2654f8c9b85cd2c8f808`

recuero committed:
Migrate files to heat_transfer

recuero committed:
Apply clang format

recuero committed:
Clean up input files and update docs.

recuero committed:
Update large media submodule

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 6ba50cc9b4de1e791ed14b6823103d4b017151a4
+Subproject commit e4812f5b9245831473dd269d4bfd5e8485f536b4
```
**Caution!** This contains a submodule update

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1828483/) on 0d33a53 wanted to post the following:


The following coverage requirement(s) failed:
- `heat_transfer` coverage rate 88.83% is less than the required 90.0%

recuero commented:
@lindsayad

maxnezdyur reviewed (approved):
All set on my end; great job!

recuero commented:
The message in https://github.com/idaholab/moose/pull/25603#issuecomment-1767446278 is requesting more coverage in the `heat_transfer` module than in its current state (for this PR is 100%). This has already been resolved (thanks @loganharbour) so the message can be ignored.
